File: apply-integration.js
#!/usr/bin/env node
/**
 * Kolbo Studio - Trimmer Integration Automation Script
 * Applies all necessary changes to integrate the trimmer feature
 */

const fs = require('fs');
const path = require('path');

const baseDir = 'G:\\Projects\\Kolbo.AI\\github\\kolbo-desktop';

function readFile(filePath) {
    return fs.readFileSync(filePath, 'utf-8');
}

function writeFile(filePath, content) {
    fs.writeFileSync(filePath, content, 'utf-8');
}

function updateIndexHtml() {
    console.log('[1/2] Updating index.html...');

    const indexPath = path.join(baseDir, 'src', 'renderer', 'index.html');
    let content = readFile(indexPath);
    let modified = false;

    // Add trimmer CSS
    if (!content.includes('trimmer.css')) {
        console.log('  - Adding trimmer.css link...');
        content = content.replace(
            /(<!-- Format Factory Styles -->\s*<link rel="stylesheet" href="css\/format-factory\.css\?v=1">)/,
            '$1\n  <!-- Trimmer Styles -->\n  <link rel="stylesheet" href="css/trimmer.css?v=1">'
        );
        console.log('  ✓ Added trimmer.css');
        modified = true;
    } else {
        console.log('  ✓ trimmer.css already present');
    }

    // Add trimmer scripts
    if (!content.includes('trimmer/video-trimmer.js')) {
        console.log('  - Adding trimmer scripts...');
        content = content.replace(
            /(<script src="js\/tab-manager\.js\?v=5"><\/script>)/,
            '$1\n  <!-- Trimmer Components (load BEFORE format-factory-manager) -->\n  <script src="js/trimmer/video-trimmer.js?v=1"></script>\n  <script src="js/trimmer/audio-trimmer.js?v=1"></script>\n  <script src="js/trimmer/trimmer-modal.js?v=1"></script>'
        );
        console.log('  ✓ Added trimmer scripts');
        modified = true;
    } else {
        console.log('  ✓ Trimmer scripts already present');
    }

    if (modified) {
        writeFile(indexPath, content);
        console.log('✓ index.html updated successfully!\n');
    } else {
        console.log('✓ index.html already up to date!\n');
    }

    return modified;
}

function updateFfmpegHandler() {
    console.log('[2/2] Updating ffmpeg-handler.js...');

    const ffmpegPath = path.join(baseDir, 'src', 'main', 'ffmpeg-handler.js');
    let content = readFile(ffmpegPath);
    let modified = false;

    // Change 1: Add trimStart and trimEnd to destructuring
    if (!content.includes('trimStart, trimEnd')) {
        console.log('  - Adding trim parameters to job destructuring...');
        content = content.replace(
            /const { id, filePath, outputFormat, outputType, settings, outputFolder } = job;/,
            'const { id, filePath, outputFormat, outputType, settings, outputFolder, trimStart, trimEnd } = job;'
        );
        console.log('  ✓ Added trim parameters');
        modified = true;
    } else {
        console.log('  ✓ Trim parameters already added');
    }

    // Change 2: Add trim to logging
    if (!content.includes('trim:')) {
        console.log('  - Adding trim info to logging...');
        const oldLog = /console\.log\('\[FFmpeg Handler\] Starting conversion:',\s*{[\s\S]*?type: outputType[\s\S]*?}\);/;
        const newLog = `console.log('[FFmpeg Handler] Starting conversion:', {
      id,
      input: filePath,
      format: outputFormat,
      type: outputType,
      trim: trimStart !== undefined ? \`\${trimStart}s - \${trimEnd}s\` : 'none'
    });`;
        content = content.replace(oldLog, newLog);
        console.log('  ✓ Added trim logging');
        modified = true;
    } else {
        console.log('  ✓ Trim logging already added');
    }

    // Change 3: Add trim logic before codec settings
    if (!content.includes('Apply trim settings')) {
        console.log('  - Adding trim processing logic...');
        content = content.replace(
            /(const command = ffmpeg\(filePath\);)\s*(\/\/ Apply conversion settings)/,
            `$1

        // Apply trim settings if specified (MUST be set before codec settings)
        if (trimStart !== undefined && trimEnd !== undefined) {
          console.log(\`[FFmpeg Handler] Applying trim: \${trimStart}s to \${trimEnd}s\`);
          // -ss: start time, -to: end time (both in seconds)
          command.setStartTime(trimStart);
          command.duration(trimEnd - trimStart);
        }

        $2`
        );
        console.log('  ✓ Added trim processing');
        modified = true;
    } else {
        console.log('  ✓ Trim processing already added');
    }

    if (modified) {
        writeFile(ffmpegPath, content);
        console.log('✓ ffmpeg-handler.js updated successfully!\n');
    } else {
        console.log('✓ ffmpeg-handler.js already up to date!\n');
    }

    return modified;
}

function main() {
    console.log('==================================');
    console.log('Kolbo Studio Trimmer Integration');
    console.log('==================================\n');

    if (!fs.existsSync(baseDir)) {
        console.error(`ERROR: Directory not found: ${baseDir}`);
        process.exit(1);
    }

    try {
        const htmlModified = updateIndexHtml();
        const ffmpegModified = updateFfmpegHandler();

        console.log('==================================');
        if (htmlModified || ffmpegModified) {
            console.log('✓ Integration Applied Successfully!');
        } else {
            console.log('✓ Already Integrated!');
        }
        console.log('==================================\n');

        console.log('⚠️  MANUAL STEP REQUIRED:');
        console.log('  Update src/renderer/js/format-factory-manager.js');
        console.log('  Follow: TRIMMER_INTEGRATION_GUIDE.md (Step 3)\n');

        console.log('Next steps:');
        console.log('1. Complete format-factory-manager.js updates');
        console.log('2. Test with: npm start');
        console.log('3. Try trimming a video or audio file!\n');

    } catch (error) {
        console.error(`\nERROR: ${error.message}`);
        process.exit(1);
    }
}

main();

File: apply_integration.py
#!/usr/bin/env python3
"""
Kolbo Studio - Trimmer Integration Automation Script
Applies all necessary changes to integrate the trimmer feature
"""

import os
import re
import sys

def read_file(path):
    """Read file content"""
    with open(path, 'r', encoding='utf-8') as f:
        return f.read()

def write_file(path, content):
    """Write file content"""
    with open(path, 'w', encoding='utf-8', newline='') as f:
        f.write(content)

def update_index_html(base_dir):
    """Update index.html to include trimmer CSS and scripts"""
    print("[1/2] Updating index.html...")

    index_path = os.path.join(base_dir, 'src', 'renderer', 'index.html')
    content = read_file(index_path)

    modified = False

    # Add trimmer CSS
    if 'trimmer.css' not in content:
        print("  - Adding trimmer.css link...")
        content = re.sub(
            r'(<!-- Format Factory Styles -->\s*<link rel="stylesheet" href="css/format-factory\.css\?v=1">)',
            r'\1\n  <!-- Trimmer Styles -->\n  <link rel="stylesheet" href="css/trimmer.css?v=1">',
            content
        )
        print("  ✓ Added trimmer.css")
        modified = True
    else:
        print("  ✓ trimmer.css already present")

    # Add trimmer scripts
    if 'trimmer/video-trimmer.js' not in content:
        print("  - Adding trimmer scripts...")
        content = re.sub(
            r'(<script src="js/tab-manager\.js\?v=5"></script>)',
            r'\1\n  <!-- Trimmer Components (load BEFORE format-factory-manager) -->\n  <script src="js/trimmer/video-trimmer.js?v=1"></script>\n  <script src="js/trimmer/audio-trimmer.js?v=1"></script>\n  <script src="js/trimmer/trimmer-modal.js?v=1"></script>',
            content
        )
        print("  ✓ Added trimmer scripts")
        modified = True
    else:
        print("  ✓ Trimmer scripts already present")

    if modified:
        write_file(index_path, content)
        print("✓ index.html updated successfully!\n")
    else:
        print("✓ index.html already up to date!\n")

    return modified

def update_ffmpeg_handler(base_dir):
    """Update ffmpeg-handler.js to support trimming"""
    print("[2/2] Updating ffmpeg-handler.js...")

    ffmpeg_path = os.path.join(base_dir, 'src', 'main', 'ffmpeg-handler.js')
    content = read_file(ffmpeg_path)

    modified = False

    # Change 1: Add trimStart and trimEnd to destructuring
    if 'trimStart, trimEnd' not in content:
        print("  - Adding trim parameters to job destructuring...")
        content = re.sub(
            r'const \{ id, filePath, outputFormat, outputType, settings, outputFolder \} = job;',
            r'const { id, filePath, outputFormat, outputType, settings, outputFolder, trimStart, trimEnd } = job;',
            content
        )
        print("  ✓ Added trim parameters")
        modified = True
    else:
        print("  ✓ Trim parameters already added")

    # Change 2: Add trim to logging
    if 'trim:' not in content:
        print("  - Adding trim info to logging...")
        old_log = r"console\.log\('\[FFmpeg Handler\] Starting conversion:', \{\s*id,\s*input: filePath,\s*format: outputFormat,\s*type: outputType\s*\}\);"
        new_log = """console.log('[FFmpeg Handler] Starting conversion:', {
      id,
      input: filePath,
      format: outputFormat,
      type: outputType,
      trim: trimStart !== undefined ? `${trimStart}s - ${trimEnd}s` : 'none'
    });"""
        content = re.sub(old_log, new_log, content, flags=re.DOTALL)
        print("  ✓ Added trim logging")
        modified = True
    else:
        print("  ✓ Trim logging already added")

    # Change 3: Add trim logic before codec settings
    if 'Apply trim settings' not in content:
        print("  - Adding trim processing logic...")
        old_code = r'(const command = ffmpeg\(filePath\);)\s*(// Apply conversion settings)'
        new_code = r'''\1

        // Apply trim settings if specified (MUST be set before codec settings)
        if (trimStart !== undefined && trimEnd !== undefined) {
          console.log(`[FFmpeg Handler] Applying trim: ${trimStart}s to ${trimEnd}s`);
          // -ss: start time, -to: end time (both in seconds)
          command.setStartTime(trimStart);
          command.duration(trimEnd - trimStart);
        }

        \2'''
        content = re.sub(old_code, new_code, content)
        print("  ✓ Added trim processing")
        modified = True
    else:
        print("  ✓ Trim processing already added")

    if modified:
        write_file(ffmpeg_path, content)
        print("✓ ffmpeg-handler.js updated successfully!\n")
    else:
        print("✓ ffmpeg-handler.js already up to date!\n")

    return modified

def main():
    print("==================================")
    print("Kolbo Studio Trimmer Integration")
    print("==================================\n")

    base_dir = r'G:\Projects\Kolbo.AI\github\kolbo-desktop'

    if not os.path.exists(base_dir):
        print(f"ERROR: Directory not found: {base_dir}")
        sys.exit(1)

    try:
        html_modified = update_index_html(base_dir)
        ffmpeg_modified = update_ffmpeg_handler(base_dir)

        print("==================================")
        if html_modified or ffmpeg_modified:
            print("✓ Integration Applied Successfully!")
        else:
            print("✓ Already Integrated!")
        print("==================================\n")

        print("⚠️  MANUAL STEP REQUIRED:")
        print("  Update src/renderer/js/format-factory-manager.js")
        print("  Follow: TRIMMER_INTEGRATION_GUIDE.md (Step 3)\n")

        print("Next steps:")
        print("1. Complete format-factory-manager.js updates")
        print("2. Test with: npm start")
        print("3. Try trimming a video or audio file!\n")

    except Exception as e:
        print(f"\nERROR: {e}")
        sys.exit(1)

if __name__ == '__main__':
    main()
